src/rvsdg/simplify_branches.rs

```diff
@@ -0,0 +1,623 @@
+            // known in the later nodes. This lets us simplify the value analysis by
+            // having empty nodes enncapsulate the information imparted by the
+            // branch.
+            for (id, edge, val) in branch_meta
+                .branches
+                .iter()
+                .flat_map(|(id, edges)| edges.iter().map(move |(edge, val)| (id, *edge, val)))
+            {
+                let Some(lit) = to_lit(val) else {
+                    continue;
+                };
+                // Count downwards from usize::MAX to avoid collisions with other placeholders
+                let node_bound = usize::MAX - var_counter;
+                var_counter += 1;
+                let (source, target) = self.graph.edge_endpoints(edge).unwrap();
+                let weight = self.graph.remove_edge(edge).unwrap();
+                let block_name = BlockName::Placeholder(node_bound);
+                let mid = self.graph.add_node(BasicBlock::empty(block_name));
+                self.graph.add_edge(source, mid, weight);
+                // NB: We rely on the optimize_direct_jumps pass to collapse this
+                // back down.
+                self.graph.add_edge(
+                    mid,
+                    target,
+                    Branch {
+                        op: BranchOp::Jmp,
+                        pos: None,
+                    },
+                );
+                val_analysis.add_assignment(mid, id.clone(), ValueInfo::Known(lit));
+            }
+            // Step 3: Compute the fixpoint of the value analysis.
+            val_analysis.compute_fixpoint(self);
+            // Step 4: Rewrite branches:
+            // * For each administrative node `n``...
+            // * For each outgoing branch [edge e1] with cond val `v` for `id`
+            // * Check if `id` was written to in `n`, if it was, then move on
+            // _unless_ we know the value of `id`; in which case we can replace the branch with a jump.
+            // * Otherwise, check if a predecessor [via edge e2] node has `v` as a
+            // known value for `id`.
+            // * If so, copy the contents of the admin node to that predecessor, and
+            // reroute e2 to the target of e1.
+            let mut scratch = Vec::new();
+            for admin_node in &branch_meta.admin_nodes {
+                let mut walker = self
+                    .graph
+                    .neighbors_directed(*admin_node, Direction::Outgoing)
+                    .detach();
+                // Don't reroute past the exit node. We want to make sure it stays reachable.
+                if admin_node == &self.exit {
+                    continue;
+                }
+                while let Some((outgoing, succ)) = walker.next(&self.graph) {
+                    let BranchOp::Cond { arg, val, .. } = self.graph[outgoing].op.clone() else {
+                        continue;
```

Comment:
All the continues and breaks make this code a little brittle to refactor
Is there a way to combine some of the conditions into boolean / matches! statements?

Reply:
In what way is it brittle? The continues and breaks are just a way of doing a filter here, and I think there's basically only one order in which to apply them all (e.g. if we refactored how condition values were represented, the right things in this code will fail to typecheck).

Reply:
It's brittle to refactors, maybe it's a me thing but I often introduce bugs when I change code with continues in it.
If it's the only way that's fine